### ISSUE TITLE
Add abandon method to Span

### ISSUE THREAD

**Issue by sirtyro** [openzipkin/brave#442]
We've run into an interesting scenario whereby we need to do span.abandon() as we have our own custom instrumentation around async messaging. So we start a span before we find out that its a duplicate. We don't want to finish or flush it as we could get up to four messages like this, which is a lot of noise. We could add a tag and then extend the reporter to simply not report spans that have a duplicate tag but it feels a bit hacky

Could we get a span.abandon() implementation?

**Comment by adriancole** [comment]
cc @felixbarny I think you have a use case for late-dropping spans when the app is a monolith (ex in this case it is more an after-the-fact sampling decision)

**Comment by adriancole** [comment]
So, this is very much like a local, after-the-fact sampling decision raised earlier in stagemonitor https://github.com/stagemonitor/stagemonitor/issues/268

`span.abandon()` means dump what you have and finish, but runs a resurrection risk.

`span.unsample()` means don't report this span (and also don't propagate it), but not necessarily finish.

`span.unsample()` could return a noopspan with the sampling flag unset in the trace context. In impl, it would write a field to MutableSpan used in the recorder so it remembers to not report it (even on GC).

**Comment by sirtyro** [comment]
I could accept that it is an after-the-fact sampling decision.

`span.unsample()` returning a noop span is confusing but i otherwise agree with what you said, i like that it would not be reported on GC either.

**Comment by adriancole** [comment]
note: I chose `unsample()` for option 2 as opposed to `sample(boolean)` as data is thrown out or not recorded at all when the span starts unsampled. late unsampling is viable, but late sampling would create junk.

**Comment by sirtyro** [comment]
I prefer unsample() as it feels more terminal, which it should. And if it returns a noop span that then helps to stop propagation?

**Comment by adriancole** [comment]
@sirtyro another choice could be to name it `span.setNoop()` which is effectively the same and might make sense as to why the result would be a noop span.

**Comment by sirtyro** [comment]
Could be confusing then as to why there is no undo, ie setReal() ? I'm ok with unsample. That is clearer than setNoop to me. But ultimately, I don't mind too much.

At the end of the day, the javadoc should make it clear why things are happening.

**Comment by adriancole** [comment]
in-process you can propagate the noop span to prevent expensive tags from being recorded or prevent new traces from being made. If you happen to be in a terminal part of code, you wouldn't propagate it anyway.

**Comment by sirtyro** [comment]
Yep cool, concur with that.

Happy with unsample or setNoop

Did you want a PR for this?

**Comment by adriancole** [comment]
The problem with setReal is data loss. tags and things like IP addresses etc are not recorded by default by most instrumentation. So, if you setReal, you'd end up with a nameless span with no information or start time. That's the problem.

**Comment by adriancole** [comment]
@sirtyro thx for the offer. Maybe `unsample()` and then if anyone opposes they can in the PR. Naming usually changes anyway before merge.

**Comment by sirtyro** [comment]
Oh, I'm not clear. I think setReal is a terrible idea. My point was that setNoop kind of implies that maybe you can undo that.

Hence, unsample feels more terminal than setNoop

**Comment by adriancole** [comment]
TL;DR; I think we should go with `abandon()` unless @felixbarny thinks `unsample()` would actually fit the stagemonitor use case. I'd go ahead with a PR on this impl for now.

Here's the latest thinking here.
* `abandon()` is simple and honest
* `unsample()` is half of a late-sampling impl due to routing data not being parsed due to guards on `span.isNoop()`

I feel like the referenced issue in stagemonitor might be a problem with the way apis currently work (ex by guarding on noop before doing things). IOTW, late local sampling is dodgy until such time as everything is always recorded locally (like happens in census).

I don't like to implement multiple things with the same purpose, so if we don't need `unsample` or it doesn't fit the stagemonitor use case, we should probably do `abandon`, as it is less code and side-effects to worry with.

**Comment by adriancole** [comment]
Long chat with @felixbarny 

So stagemonitor supports span-level sampling, for example to help control scenarios where a bug results in an excess of spans:

https://github.com/stagemonitor/stagemonitor/wiki/Configuration-Options#sampling-probability-for-spans-in--per-operation-type

Currently, Brave (and de-facto zipkin for that matter) conflate sampled status with recording status for the entire trace. I've opened #444 to entertain the idea of changing this in the future. In the mean time, accomplishing span-tiered unsampling can be something we do in brave-opentracing (as there's no span-scoped sample option in brave!). I've offered to do that work.

**Comment by sirtyro** [comment]
This could be something we could also find useful. I think we have a process that spawns hundreds or thousands of rabbit messages in a batch. I'm keen to see how zipkin and brave handle that :) If it proves to be too much, then the sampling you described may help.
Also, @garyd203 managed to crash chrome with the zipkin UI yesterday. I think he had a trace with over 3,000 spans and the UI and chrome struggled then died.

**Comment by sirtyro** [comment]
I should add though, that was the result of a bug on his part. It wasn't expected!

**Comment by adriancole** [comment]
On operation in a trace sampling. One thing thet prevented us in the past
from doing this was that there is no way to tell the trace is missing data.
In census I think they have a dropped event counter on a span, which could
hint that you are looking at partial data (which helps from the confusion
trying to figure out why there is latency but not many ops)

**Comment by adriancole** [comment]
On UI choking on thousands of spans, yeah thats missing a performance
person. In zipkin theres a ticket but needs hands to tune the javascript
(which has never been tuned)

**Comment by felixbarny** [comment]
Stagemonitor also adds the number of external requests and their total
execution time to the span. Dropped counter sounds useful as well!

Adrian Cole <notifications@github.com> schrieb am Fr., 16. Juni 2017 03:32:

**Comment by adriancole** [comment]
closing on abandon() impl. We may revisit data policy and supporting changes in the model, but a separate issue I think (motivated by this)
